lib/radiator_web/controllers/admin/network_controller.ex

```diff
@@ -81,4 +90,66 @@ defmodule RadiatorWeb.Admin.NetworkController do
         other
     end
   end
+
+  def add_collaborator(conn, %{"id" => id, "collaborator" => collaborator}) do
```

Comment:
I try to keep controllers RESTful, even in CMS UIs. Here, the resource you're modifying is a "network collaboration". So I would have a `NetworkCollaborationController` (or generic `CollaborationController`) with `create` and `delete` implementation instead of `add_collaborator` and `remove_collaborator` here. It has the benefit of similar implementations/structure of CMS controllers and actual REST API controllers. It's encouraged by Phoenix through the [resources](https://hexdocs.pm/phoenix/1.4.8/Phoenix.Router.html#resources/4) Router Macro.

See also https://hexdocs.pm/phoenix/routing.html#resources

Reply:
How would one adopt this for networks, podcasts and episodes alike? are these 3 different controller resources living under /:resourcepath/collaborator?

Reply:
I'll try. But I'm slightly confused why we still do have parallel controller hierarchies in API and UI then. E.g. should't that then just be switched based on the pipeline/resulttype vs. having parallel hierarchies?

Reply:
Mutations in particular are probably very similar in API and UI, true. Not aware of a way to bring these together but open to suggestions. Reading endpoints (index, show) though will often be different as the API has a narrow focus while the UI may display many different things at once.

To answer your first question, you can use resource nesting:

```elixir
resources "/networks", NetworkController do
  resources "/collaboration", CollaborationController, only: [:create, :delete]
end

resources "/podcasts", NetworkController do
  resources "/collaboration", CollaborationController, only: [:create, :delete]
end

# ...
```

That generates routes like `DELETE  /networks/:network_id/collaborations/:id`

In that case you can have one generic `CollaborationController` and match on the params, as one will have a `network_id` in the request and the other one `podcast_id`. etc. `mix phx.routes` is helpful to see what routes witch what paramters are generated.